deploy/crds/apps_v1alpha1_apicast_crd.yaml

```diff
@@ -80,6 +80,15 @@ spec:
             serviceAccount:
               type: string
           type: object
+          anyOf:
```

Comment:
This AnyOf is applied in the "Spec" section for all the fields in them.

Doing this makes not possible to add validation for other set of fields.
For example if we were to have two other fields where we wanted to make sure one of the two (or the two) had to be set and we add it in this anyOf then it would have 4 elements.
That's not what we intend.

In any case we should move the adminPortal and the embeddedConfig fields into a new section of the spec (for example configurationSource) and there define the two fields and add the anyOf in that section.

Also we should talk whether we want to do this. There are some situations that might happen in the future that would break this validation. For example if you have three fields A,B,C and A and B are compatible but A and C are not. In that scenario AnyOf would not work because the combination A,C is not compatible (this might be an edge case but can theoretically happen)

Reply:
That is not true.

`anyOf` is a conditional to apply subschemas. It mean, any of the following must happen, otherwise validation fails. it does NOT check ALL the attributes.

Besides, there is no need (unless there is another reason to do) to group in a new section, structural schemas (based on openapi 3 validation) is flexible enough. For instance, you may have another conditional subschema to say, for instance, `anyOf` `image` or `logLevel` should be present like this:

```
anyOf:                                            
  - properties:                                   
      adminPortalCredentialsRef:                  
        type: object                              
    required: ["adminPortalCredentialsRef"]       
  - properties:                                   
      embeddedConfigurationSecretRef:             
        type: object                              
    required: ["embeddedConfigurationSecretRef"]  
anyOf:                                            
  - properties:                                   
      image:                                      
        type: string                              
    required: ["image"]                           
  - properties:                                   
      logLevel:                                   
        type: string                              
    required: ["logLevel"]                        
```

Anyway, having this enforcement at CRD level (cluster level) and not at operator level is something worth to be discussed in other PR. I will open another PR enforcing any of the following configuration sources to exist, `adminPortalCredentialsRef` or `embeddedConfigurationSecretRef`

Reply:
https://github.com/3scale/apicast-operator/pull/10

Reply:
As far as I know It DOES check all attributes. Between all attributes in spec it checks that any of the specified ones in AnyOf are specified in spec.

Regarding specifying multiple times anyOf :+1: if you have seen we could do this then if we decide to do that would be a good option :+1: . Have you found this behavior described in the documentation somewhere?

Reply:
Conditions check for the specified fields on the current context. It does NOT check all attributes. That is why conditions expect a list of properties.

From [Openapi Docs](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/). The hint that made me try was the text in the doc `validate a value against multiple criteria.`

Reply:
But you have to check those conditions from a set of fields. In this case, it is checked that among all the fields in the spec section any of them (1 or more) are the ones specified in the anyOf section